### ISSUE TITLE
Load subfolders

### ISSUE THREAD

**Issue by Eomm** [fastify/fastify-autoload#51]
## 🚀 Feature Proposal

Given the tree:

* folder
  * file a
  * folder lv 1
    * file b
    * folder lv 2
       * file c

Only the `file a` is loaded and the other directories are ignored

In this example, I would `register` every folder and put all the files in the same context.

## Motivation

This lets the user organize the application in a more flexible structure where the URLs (except some prefix-override) is mapped to the project tree

## Example

```
fastify.register(require('fastify-autoload'), {
  dir: x,
  recursive: true // as rm -rf 😅
})
```

**Comment by mcollina** [comment]
I like this proposal! However I would automatically add prefixes for these. However, I'm not sure how it would all work.. because more often than not there are intermediate parameters when nesting directories, i.e. `/books/42/author/24/`.
How are Ruby on Rails or Django implementing this very same feature?

**Comment by Eomm** [comment]
I'm asking around to the information.

Right now I have 2 ideas:

- a `index.js` file that is loading first and expose a `module.export.autoload.prefix = '/my/:custom'`
- an opt `prefixController: function(dirName)` that is called foreach directory and the user should return:
  - a string to use as prefix
  - falso to dont' use the dirname as prefix
  - true do use the dirname as prefix

The idea is:
each directory is a context and all files in that directory are in the same context

**Comment by mcollina** [comment]
How about we add a well known js or json file in every folder to customize
certain behaviors? As an example, we could use this place to register
per-directory hooks or decorators.
This file could well be index.js, but we might need some extra properties
attached.

Il giorno sab 31 ago 2019 alle 15:29 Manuel Spigolon <
notifications@github.com> ha scritto:

**Comment by Eomm** [comment]
What about `autoload.json` with this format?

```
{
  "prefix": "/:userId", /** the prefix to use instead of the directory name*/
  "logLevel": "debug", /** the log level to apply */
  "skip": true, /** skip the loading of this directory and all the content (files and subfolders)*/
}
```

**Comment by mcollina** [comment]
+1!

**Comment by francisbrito** [comment]
Is it OK if I have a look at this one @mcollina @Eomm ?

**Comment by Eomm** [comment]
Go for it 💪

Take care that right now the file loading is parallel:
https://github.com/fastify/fastify-autoload/blob/master/index.js#L35

Instead, to archive this feature is necessary to (pseudo-code):

1. create a context and start looping the input `dir` content (`.register`)
1. if it is a file, register it to the parent scope (`.register(fastity-plugin(file))`)
1. if it is a directory execute the first step

If you have question don't hesitate to ask!

**Comment by SkeLLLa** [comment]
I've implemented something similar: https://github.com/SkeLLLa/fastify-autoload-recursive.
Feel free to reuse the code if you'd like.

And here some notes that I've found during implementing it: 

Sometimes it's useful to define file order in which they are registered. For example if we have plugin that adds shared shema and plugin that adds route in the same folder, then schema should be registered before. So that's why I created 3 patterns for files: `schema`, `options`, `ignore`.
Options file contains a function that may return additional options (that could be calculated dynamically, so you couldn't pass them directly) and after that they are merged with static options and passed to `register` function. Such files evaluated in first place.
Schema files contain shared schemas that are used later. Such files are registered before other files, so later we can use shared shemas.
Ignore patter is used to ignore files, for example, some helpers with common code that are used only in specific plugin or service. Those are ignored by autoloader and required directly in other files.

I don't know if `schema` and `options` settings are useful for others, but I think that ignore pattern here is a must have feature for recursive loading.

**Comment by alemagio** [comment]
Hi, is anyone working on this?

**Comment by mcollina** [comment]
Not that I think of.

**Comment by alemagio** [comment]
If you don't mind I'll try to do it.

**Comment by mcollina** [comment]
Go ahead: This should build on the amazing work done in https://github.com/fastify/fastify-autoload/pull/33.

**Comment by alemagio** [comment]
Hi, just letting you know I'm not dead :sweat_smile: 
I'm working on it and I'll give you a draft in the next few days (hopefully) with a unit test and a basic implementation just to be sure I'm on the right path and to have feedback.

**Comment by mcollina** [comment]
This was implemented as part of #76.
